test/DepositAbstract.t.sol

```diff
@@ -0,0 +1,19 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.13;
+
+import "./ZeroState.t.sol";
+
+abstract contract DepositAbstract is ZeroState {
+     
+     function setUp() public override virtual {
+         super.setUp();
+         
+         vm.prank(sonic);
+         rings.approve(address(vault), 100000);
```

Comment:
i guess its ok to put approve here, but id rather see the approve right before the trasnfer.  it makes it more readable.  someone reading the test might not realize its been approved in the test setup.

but there's a bigger problem and thats with the test setup and the different abstract contracts and state layers.  i know it can be confusing and its also a wip so def open to feedback.  but your state abstract contracts seem to be lagging one step.  all the stuff i see in DepositAbstract belongs in ZeroState (but if we are removing approve from setup then that means this logic gets deleted altogether). 

But then the logic you have in the Withdraw setup, i think belongs in the Deposit setup.  It's probably better to think  of it in past tense.  The "deposited" state means funds have already been deposited -- its not the setup u need to before doing a deposit.

example:
ZeroState -- no state changes have happened yet.  here we can set up users and possibly other things that will be used by all states.

Then do some tests on zerostate.  
First test negation (like can't withdraw if you don't have a balance).
Then test things u can do from the zero state (like deposit).

Once you've proven in a test that you can deposit, then you can add that to the Deposited state.  So you can put the logic that was just found in the deposit test into the setup for the deposited state...and so on.

Does this help make sense?  Please feel free to reach out here or in slack if it didn't.  This is all new ground for us and we are doing our best to develop new best practices and patterns.

Reply:
made changes and renamed some contracts to describe the state it creates before a test